internal/presentation/presentation.go

```diff
@@ -126,6 +131,118 @@ func (e Output) undefined() bool {
 	return len(e.Result) == 0 && (e.Partial == nil || len(e.Partial.Queries) == 0)
 }
 
+// NewOutputErrors creates a new slice of OutputError's based
+// on the type of error passed in. Known structured types will
+// be translated as appropriate, while unknown errors are
+// placed into a structured format with their string value.
+func NewOutputErrors(err error) []OutputError {
+	var errs []OutputError
+	if err != nil {
+		// Handle known structured errors
+
+		switch typedErr := err.(type) {
+		case *ast.Error:
+			oe := OutputError{
+				Type:    "ast",
+				Code:    typedErr.Code,
+				Message: typedErr.Message,
+				Details: typedErr.Details,
+				err:     typedErr,
+			}
+
+			// TODO(patrick-east): Why does the JSON marshaller marshal
+			// location as `null` when err.location == nil?!
+			if typedErr.Location != nil {
+				oe.Location = typedErr.Location
+			}
+			errs = []OutputError{oe}
+		case *topdown.Error:
+			errs = []OutputError{{
+				Type:     "topdown",
+				Code:     typedErr.Code,
+				Message:  typedErr.Message,
+				Location: typedErr.Location,
+				err:      typedErr,
+			}}
+		case *storage.Error:
+			errs = []OutputError{{
+				Type:    "storage",
+				Code:    typedErr.Code,
+				Message: typedErr.Message,
+				err:     typedErr,
+			}}
+
+		// The cases below are wrappers for other errors, format errors
+		// recursively on them.
+		case ast.Errors:
+			for _, e := range typedErr {
+				if e != nil {
+					errs = append(errs, NewOutputErrors(e)...)
+				}
+			}
+		case rego.Errors:
+			for _, e := range typedErr {
+				if e != nil {
+					errs = append(errs, NewOutputErrors(e)...)
+				}
+			}
+		case loader.Errors:
+			{
+				for _, e := range typedErr {
+					if e != nil {
+						errs = append(errs, NewOutputErrors(e)...)
+					}
+				}
+			}
+		default:
+			// Any errors which don't have a structure we know about
+			// are converted to their string representation only.
+			errs = []OutputError{{
+				Message: err.Error(),
+				err:     typedErr,
+			}}
+		}
+	}
+	return errs
+}
+
+// OutputErrors is a list of errors encountered
+// which are to presented.
+type OutputErrors []OutputError
+
+func (e OutputErrors) Error() string {
+	if len(e) == 0 {
+		return "no error(s)"
+	}
+
+	if len(e) == 1 {
+		return fmt.Sprintf("1 error occurred: %v", e[0].Error())
+	}
+
+	var s []string
+	for _, err := range e {
+		s = append(s, err.Error())
+	}
+
+	return fmt.Sprintf("%d errors occurred:\n%s", len(e), strings.Join(s, "\n"))
+}
+
+// OutputError provides a common structure for all OPA
+// library errors so that the JSON output given by the
+// presentation package is consistent and parsable.
+type OutputError struct {
+	Message  string      `json:"message"`
+	Type     string      `json:"type,omitempty"`
```

Comment:
@patrick-east Is this Type field necessary? We have the code field that should indicate the kind of details. I'm tempted to remove it it.

Reply:
I could go either way on it. From a parsing perspective it makes it potentially easier to key off the type to know what fields to expect.. arguably the same could be said for the code too.

If you'd prefer to remove it I'm on board, just let me know.

Reply:
My preference would be to remove it. The code should tell users what kind of details are available.

Reply:
Done